[Test] Validate metrics workload on prometheus

sayaleeraut opened the pull request:
The test case generates workload by creating PVCs
and fetching the metrics for the PVCs from prometheus.
The metric workload should not cause any issues on
prometheus and the details should be fetched properly.

srivickynesh commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,82 @@ def test_restart_prometheus_glusterfs_pod(self):
+
+    @pytest.mark.tier2
+    def test_metrics_workload_on_prometheus(self):
+        """Validate run metrics workload on prometheus"""
```
"""Validate metrics workload on prometheus"""

srivickynesh commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,82 @@ def test_restart_prometheus_glusterfs_pod(self):
+        """Validate run metrics workload on prometheus"""
+
+        self.prefix = "autotest-{}".format(utils.get_random_str())
+
```
self._registry_project_name = self.registry_sc.get(
            'restsecretnamespace')

srivickynesh replied:
For this test, take project name from config as mentioned above

adityaramteke commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,82 @@ def test_restart_prometheus_glusterfs_pod(self):
+        # Add check for CRS version
+        openshift_ops.switch_oc_project(
+            self._master, self._registry_project_name)
+        if not self.is_containerized_gluster():
+            self.skipTest("Skipping this test case as CRS"
+                          " version check can not be implemented")
```
Test case doesn't suggest executing only on converged mode setup. This check is not required.

adityaramteke commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,82 @@ def test_restart_prometheus_glusterfs_pod(self):
+            self.skipTest("Skipping this test case as CRS"
+                          " version check can not be implemented")
+
+        # Get one of the prometheus pod name and respective pvc name
```
```suggestion
        # Skip test if Prometheus pods are not present
```

adityaramteke commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,82 @@ def test_restart_prometheus_glusterfs_pod(self):
+            self.skipTest(
+                prometheus_pods, "Skipping test as prometheus"
+                " pod is not present")
+
```
```suggestion
# Get one of the Prometheus pod name and respective PVC name
```

adityaramteke commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,82 @@ def test_restart_prometheus_glusterfs_pod(self):
+                " pod is not present")
+
+        prometheus_pod = list(prometheus_pods.keys())[0]
+        pvc_name = openshift_ops.oc_get_custom_resource(
+            self._master, "pod",
+            ":.spec.volumes[*].persistentVolumeClaim.claimName",
+            prometheus_pod)
+        self.assertTrue(
+            pvc_name, "Failed to get PVC name for prometheus"
+            " pod {}".format(prometheus_pod))
+        pvc_name = pvc_name[0]
+        self.verify_iscsi_sessions_and_multipath(
+            pvc_name, prometheus_pod, rtype='pod',
+            heketi_server_url=self._registry_heketi_server_url,
+            is_registry_gluster=True)
```
```suggestion
        pvc_custom = ":.spec.volumes[*].persistentVolumeClaim.claimName"
        pvc_name = openshift_ops.oc_get_custom_resource(
            self._master, "pod", pvc_custom, prometheus_pod)
        self.assertTrue(
            pvc_name, "Failed to get PVC name for prometheus"
            " pod {}".format(prometheus_pod))
        self.verify_iscsi_sessions_and_multipath(
            pvc_name[0], prometheus_pod, rtype='pod',
            heketi_server_url=self._registry_heketi_server_url,
            is_registry_gluster=True)
```

sayaleeraut replied:
@adityaramteke ; I do not understand the purpose of adding 'pvc_custom' as a variable as we are not using it anywhere else?

adityaramteke replied:
@sayaleeraut, we are defining the custom variable just for the simplicity of the code. Please refer to one of the TC https://github.com/gluster/glusterfs-containers-tests/blob/master/tests/functional/logging/test_logging_validations.py#L115

sayaleeraut replied:
@adityaramteke; okay, I will add it.

adityaramteke commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,82 @@ def test_restart_prometheus_glusterfs_pod(self):
+        self.addCleanup(openshift_ops.switch_oc_project,
+                        self._master, self._registry_project_name)
+
+        # Create PVCs and app pods
+        pvc_count, batch_count = 5, 5
+        for _ in range(batch_count):
+            test_pvc_names = self.create_and_wait_for_pvcs(
+                pvc_size=1, pvc_name_prefix=self.prefix,
+                pvc_amount=pvc_count, sc_name=self.sc_name, timeout=600,
+                wait_step=10)
```
```suggestion
        # Create PVCs and app pods
        pvc_size, pvc_count, batch_count = 1, 5, 5
        for _ in range(batch_count):
            test_pvc_names = self.create_and_wait_for_pvcs(
                pvc_size, pvc_name_prefix=self.prefix,
                pvc_amount=pvc_count, sc_name=self.sc_name, timeout=600,
                wait_step=10)
```

vamahaja commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,78 @@ def test_restart_prometheus_glusterfs_pod(self):
+            metric='kube_persistentvolumeclaim_info')
+        self._fetch_metric_from_promtheus_pod(
+            metric='kube_pod_spec_volumes_persistentvolumeclaims_info')
+        self.addCleanup(openshift_ops.switch_oc_project,
```
Why this cleanup needed here ?

sayaleeraut replied:
If we have the cleanup at line 1050, log snippet after the test case execution is completed:
========================================================================
2021-03-08 13:35:19,324 INFO (tearDown) Ending Test: tests.functional.prometheous.test_metrics_workload_on_prometheus.TestMetricsWorkloadOnPrometheus.test_metrics_workload_on_prometheus : 13_33_08_03_2021
2021-03-08 13:35:19,324 INFO (run) root@10.70.47.18 (cp): oc project glusterfs
2021-03-08 13:35:19,810 INFO (_log_results) ^[[34;1mRETCODE (root@10.70.47.18): 0^[[0m
2021-03-08 13:35:19,811 INFO (_log_results) ^[[47;30;1mSTDOUT (root@10.70.47.18)...
Now using project "glusterfs" on server "https://dhcp47-18.lab.eng.blr.redhat.com:8443".
^[[0m
2021-03-08 13:35:19,811 INFO (run) root@10.70.47.18 (cp): oc scale  --replicas=0 dc/autotests-dc-with-app-io-2qss0lmjndssiw dc/autotests-dc-with-app-io-fdyt7a6ypxvlsu dc/autotests-dc-with-app-io-x92a30rrlxgooq dc/autotests-dc-with-app-io-wafrwrye8pz1hh dc/autotests-dc-with-app-io-lqwpl4ez2toy99
2021-03-08 13:35:20,385 INFO (_log_results) ^[[34;1mRETCODE (root@10.70.47.18): 0^[[0m
2021-03-08 13:35:20,385 INFO (_log_results) ^[[47;30;1mSTDOUT (root@10.70.47.18)...
deploymentconfig.apps.openshift.io/autotests-dc-with-app-io-2qss0lmjndssiw scaled
deploymentconfig.apps.openshift.io/autotests-dc-with-app-io-fdyt7a6ypxvlsu scaled
deploymentconfig.apps.openshift.io/autotests-dc-with-app-io-x92a30rrlxgooq scaled
deploymentconfig.apps.openshift.io/autotests-dc-with-app-io-wafrwrye8pz1hh scaled
deploymentconfig.apps.openshift.io/autotests-dc-with-app-io-lqwpl4ez2toy99 scaled
^[[0m


But if the cleanup is removed, we see following errors:
=============================================
2021-03-17 16:23:29,867 INFO (tearDown) Ending Test: tests.functional.prometheous.test_metrics_workload_on_prometheus.TestMetricsWorkloadOnPrometheus.test_metrics_workload_on_prometheus : 16_21_17_03_2021
2021-03-17 16:23:29,867 INFO (run) root@10.70.47.18 (cp): oc scale  --replicas=0 dc/autotests-dc-with-app-io-bsa6p4yfteoxrb dc/autotests-dc-with-app-io-o63h39lripebbq dc/autotests-dc-with-app-io-3rmbf3byjfkpcz dc/autotests-dc-with-app-io-gum155zb3pzaxe dc/autotests-dc-with-app-io-w7z31tnx9luep2
2021-03-17 16:23:32,433 INFO (_log_results) RETCODE (root@10.70.47.18): 1
2021-03-17 16:23:32,433 INFO (_log_results) STDERR (root@10.70.47.18)...
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-bsa6p4yfteoxrb" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-o63h39lripebbq" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-3rmbf3byjfkpcz" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-gum155zb3pzaxe" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-w7z31tnx9luep2" not found

2021-03-17 16:23:32,434 ERROR (cmd_run) Failed to execute command 'oc scale  --replicas=0 dc/autotests-dc-with-app-io-bsa6p4yfteoxrb dc/autotests-dc-with-app-io-o63h39lripebbq dc/autotests-dc-with-app-io-3rmbf3byjfkpcz dc/autotests-dc-with-app-io-gum155zb3pzaxe dc/autotests-dc-with-app-io-w7z31tnx9luep2' on '10.70.47.18' node. Got non-zero return code '1'. Err: Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-bsa6p4yfteoxrb" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-o63h39lripebbq" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-3rmbf3byjfkpcz" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-gum155zb3pzaxe" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "autotests-dc-with-app-io-w7z31tnx9luep2" not found

sayaleeraut committed:
[Test] Validate metrics workload on prometheus

The test case generates workload by creating PVCs
and fetching the metrics for the PVCs from prometheus.
The metric workload should not cause any issues on
prometheus and the details should be fetched properly.

Signed-off-by: sayaleeraut <saraut@redhat.com>

adityaramteke commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,78 @@ def test_restart_prometheus_glusterfs_pod(self):
+        """Validate metrics workload on prometheus"""
+
+        self.prefix = "autotest-{}".format(utils.get_random_str())
+        self._registry_project = self.registry_sc.get(
+            'restsecretnamespace')
```
If my requested changes from +1015 to +1051 are applicable then this can be removed. 
```
```

adityaramteke commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,78 @@ def test_restart_prometheus_glusterfs_pod(self):
+        self._fetch_metric_from_promtheus_pod(
+            metric='kube_persistentvolumeclaim_info')
+
+        # Create storage class
+        openshift_ops.switch_oc_project(
+            self._master, self._registry_project)
+        self.sc_name = self.create_storage_class(
+            vol_name_prefix=self.prefix, glusterfs_registry=True)
+        self.addCleanup(openshift_ops.switch_oc_project,
+                        self._master, self._registry_project)
+
+        # Create PVCs and app pods
+        pvc_size, pvc_count, batch_count = 1, 5, 5
+        for _ in range(batch_count):
+            test_pvc_names = self.create_and_wait_for_pvcs(
+                pvc_size, pvc_name_prefix=self.prefix,
+                pvc_amount=pvc_count, sc_name=self.sc_name, timeout=600,
+                wait_step=10)
+            self.create_dcs_with_pvc(
+                test_pvc_names, timeout=600, wait_step=5,
+                dc_name_prefix="autotests-dc-with-app-io",
+                space_to_use=1048576)
+
+        # Check from the prometheus pod for the PVC space usage
+        openshift_ops.switch_oc_project(
+            self._master, self._prometheus_project_name)
+        mount_path = "/prometheus"
+        cmd = "oc exec {0} -- df -PT {1} | grep {1}".format(
+            prometheus_pod, mount_path)
+        out = self.cmd_run(cmd)
+        self.assertTrue(out, "Failed to get info about mounted volume. "
+                             "Output is empty.")
+
+        # Try to fetch metric from prometheus pod
+        self._fetch_metric_from_promtheus_pod(
+            metric='kube_persistentvolumeclaim_info')
+        self._fetch_metric_from_promtheus_pod(
+            metric='kube_pod_spec_volumes_persistentvolumeclaims_info')
+        self.addCleanup(openshift_ops.switch_oc_project,
+                        self._master, self._registry_project)
```
Application pods and respective PVCs can be created in any namespace. We should not bound to _registry_project project. What matters the most is we should use glusterfs_registry storage_class to create app pod. Currently, you are switching between multiple projects, and you also seeing some issues while cleanup. I will suggest the following changes.


```suggestion
        # Create storage class
        self.sc_name = self.create_storage_class(
            vol_name_prefix=self.prefix, glusterfs_registry=True)

        # Create PVCs and app pods
        pvc_size, pvc_count, batch_count = 1, 5, 5
        for _ in range(batch_count):
            test_pvc_names = self.create_and_wait_for_pvcs(
                pvc_size, pvc_name_prefix=self.prefix,
                pvc_amount=pvc_count, sc_name=self.sc_name, timeout=600,
                wait_step=10)
            self.create_dcs_with_pvc(
                test_pvc_names, timeout=600, wait_step=5,
                dc_name_prefix="autotests-dc-with-app-io",
                space_to_use=1048576)

        # Check from the prometheus pod for the PVC space usage
        mount_path = "/prometheus"
        cmd = "oc exec {0} -- df -PT {1} | grep {1}".format(
            prometheus_pod, mount_path)
        out = self.cmd_run(cmd)
        self.assertTrue(out, "Failed to get info about mounted volume. "
                             "Output is empty.")

        # Try to fetch metric from prometheus pod
        self._fetch_metric_from_promtheus_pod(
            metric='kube_persistentvolumeclaim_info')
        self._fetch_metric_from_promtheus_pod(
            metric='kube_pod_spec_volumes_persistentvolumeclaims_info')
```

vamahaja commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,78 @@ def test_restart_prometheus_glusterfs_pod(self):
+            metric='kube_pod_spec_volumes_persistentvolumeclaims_info')
+        self.addCleanup(openshift_ops.switch_oc_project,
+                        self._master, self._registry_project)
+ 
```
Remove this line. PEP8 validation failing with error -
```
pep8 run-test: commands[3] | flake8 tests/functional/prometheous/test_prometheus_validations.py
tests/functional/prometheous/test_prometheus_validations.py:1052:1: W293 blank line contains whitespace

```

sayaleeraut replied:
Done

vamahaja commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,77 @@ def test_restart_prometheus_glusterfs_pod(self):
+        """Validate metrics workload on prometheus"""
+
+        self.prefix = "autotest-{}".format(utils.get_random_str())
+        self._registry_project = self.registry_sc.get(
```
Move this variable declaration after skip statement and add check for `registry_sc` key.

sayaleeraut replied:
Done

vamahaja reviewed (changes requested):
PEP8 validation failed with error -
```
pep8 run-test: commands[3] | flake8 tests/functional/prometheous/test_prometheus_validations.py
tests/functional/prometheous/test_prometheus_validations.py:1056:1: W391 blank line at end of file
ERROR: InvocationError for command /home/vamahaja/Codes/gct/glusterfs-containers-tests/.tox/pep8/bin/flake8 tests/functional/prometheous/test_prometheus_validations.py (exited with code 1)
```

SushilG96 commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,81 @@ def test_restart_prometheus_glusterfs_pod(self):
+        """Validate metrics workload on prometheus"""
+
+        # Skip test if the prometheus pods are not present
+        openshift_ops.switch_oc_project(
```
Add a cleanup for the project switch

sayaleeraut replied:
Cleanup to switch to self._registry_project ?  At line 992 ?

SushilG96 replied:
You can first get the current project where currently it is.
Add it after 985

sayaleeraut replied:
After discussing with @SushilG96 ; it was agreed upon that it is not required here, since the method self._fetch_metric_from_promtheus_pod() called at Line 1016 already does it.

SushilG96 commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,81 @@ def test_restart_prometheus_glusterfs_pod(self):
+        # Get one of the prometheus pod name and respective pvc name
+        prometheus_pod = list(prometheus_pods.keys())[0]
+        pvc_custom = ":.spec.volumes[*].persistentVolumeClaim.claimName"
+        pvc_name = openshift_ops.oc_get_custom_resource(
```
use 
```
pvc_name = openshift_ops.oc_get_custom_resource(
              self._master, "pod", pvc_custom, prometheus_pod)[0]
```
And now check for assertTrue, Now you can ignore line number 1009

sayaleeraut replied:
You mean:
        pvc_name = openshift_ops.oc_get_custom_resource(
            self._master, "pod", pvc_custom, prometheus_pod)[0[
        self.assertTrue(
            pvc_name, "Failed to get PVC name for prometheus"
            " pod {}".format(prometheus_pod))
        self.verify_iscsi_sessions_and_multipath(
            pvc_name, prometheus_pod, rtype='pod',
            heketi_server_url=self._registry_heketi_server_url,
            is_registry_gluster=True)

SushilG96 replied:
Right.

sayaleeraut replied:
Done

SushilG96 commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,81 @@ def test_restart_prometheus_glusterfs_pod(self):
+            self._master, self._registry_project)
+        self.sc_name = self.create_storage_class(
+            vol_name_prefix=self.prefix, glusterfs_registry=True)
+        self.addCleanup(openshift_ops.switch_oc_project,
```
Why switching back to `self._registry_project` as you are switching to `self._registry_project`  in line number 1020, So the clean will be to the project from where it was previously i.e before switching to `self._registry_project`

sayaleeraut replied:
You are suggesting we do not need the cleanup here for switching to self._registry_project ?

SushilG96 replied:
No, I mean add a check to know in which project currently you are, i.e before switching and then in clean up use the project to switch back

vamahaja replied:
@SushilG96 in this step storage class is created in `self._registry_project` and cleanup to delete same sc is added in the function without mention of namespace. But in this TC at few steps namespace is changed to fetch metrics from prometheus, which will cause issue while executing the cleanup to delete sc created in `self._registry_project`. To avoid this situation we are adding cleanup to move `self._registry_project`  namespace.

sayaleeraut replied:
Using 'self._registry_project' in addCleanup() at Line 1024.

SushilG96 commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,81 @@ def test_restart_prometheus_glusterfs_pod(self):
+                prometheus_pods, "Skipping test as prometheus"
+                " pod is not present")
+
+        self.prefix = "autotest-{}".format(utils.get_random_str())
```
Move this line after the check for `self.registry_sc`

sayaleeraut replied:
Done

SushilG96 commented on tests/functional/prometheous/test_prometheus_validations.py:
```diff
@@ -974,3 +975,81 @@ def test_restart_prometheus_glusterfs_pod(self):
+            metric='kube_persistentvolumeclaim_info')
+        self._fetch_metric_from_promtheus_pod(
+            metric='kube_pod_spec_volumes_persistentvolumeclaims_info')
+        self.addCleanup(openshift_ops.switch_oc_project,
```
Plz have a check before line number 1040 so that we are currently at i.e which project and use that in the cleanup.

sayaleeraut replied:
Can you please refer to this comment https://github.com/gluster/glusterfs-containers-tests/pull/17#discussion_r595902137 and let me know if we still need to add a check before line 1040 like you suggested?

SushilG96 replied:
Yes, that okay
For cleanup you need to get back to the project from where you were i.e before switching as in 1040, so in clean you need to move there

sayaleeraut replied:
After discussing with @vamahaja ; as issues were seen in TC clean-up steps (if switch to 'current_project' was done), decided to use 'self._registry_project' itself in the addCleanup() at Line 1055

SushilG96 reviewed (approved):
LGTM
